### ISSUE TITLE
Question: Is it possible to filter a feed before it is processed by elfeed?

### ISSUE THREAD

**Issue by herrsimon** [skeeto/elfeed#339]
Hello,

I need to monitor several scientific journals from a certain publisher and unfortunately their article feeds are broken in the sense that whenever a new article is accepted for submission, the feed delivers all articles from the current journal issue with a new guid and pubDate, resulting in a lot of duplicate entries (the same article appears on average about ten times on the feed). Fixing the feed is very easy, as the guid is of the form $id_$date, where $id is the article id (which never changes), so only the _$date part needs to be removed. As I would like to avoid setting up a local webserver which proxies the feed and pipes it through a script before delivering: Is there some hook in elfeed so that after the feed is fetched by curl it can be modified before elfeed processes it?

PS: The simple solution, i.e. asking the journal to fix their feed is unfortunately not working, I tried many times and never got an answer. 

PPS: Here is one of the feeds in question https://projecteuclid.org/feeds/euclid.ejp_article_rss.xml

**Comment by herrsimon** [comment]
As far as I understand after looking at elfeed-curl.el, I could simply change the value of  elfeed-curl-program-name to some wrapper script which does the necessary adjustments, correct?

**Comment by skeeto** [comment]
The closest thing Elfeed has is elfeed-new-entry-hook which is called on 
new entries when they are discovered. However, one of the things you 
can't change in that hook is the entry ID. It's a bit of a chicken and 
egg problem in that the reason it's considered new is because it has an 
ID that wasn't seen before.

In your case what you'd really need is a way is for the hook to reject 
the new entry. I originally made the new entry hook for tags, but it 
really should be more flexible and powerful than it is. Maybe you should 
be able to change literally anything you want about the entry, even down 
to rejecting it with a special return value (:reject keyword?).

Your idea with wrapping curl could technically work, but I don't 
recommend it. There's essentially an undocumented protocol between curl 
and Elfeed that occurs via the passed command line arguments. Your 
middleware would need to understand this protocol to manipulate the 
contents in-between. IMHO, if you go this route, you'd be much better 
off hosting your own server that does the manipulation itself.

So, in summary, there's not really a good solution within Elfeed, but I 
wish elfeed-new-entry-hook was a little more powerful so that it could 
be that solution (i.e. a TODO list item).

**Comment by herrsimon** [comment]
Thanks a lot for this detailed answer, I agree that a more powerful `elfeed-new-entry-hook` would be a nice solution, but even then something like an `elfeed-post-fetch-hook` which passes the feed curl has fetched to every function in the hook and expects valid rss/atom would make sense as well, also in terms of efficiency regarding my specific use case. Going one step further, how about supporting to fetch feeds from a local file as well, i.e. feeds of the form `file://path/to/feed.xml`? Both of these functionalities should be pretty quick to implement and I would be happy to try to do it if you don't have the time (my lisp-knowledge is very limited but probably sufficient for these changes).

Also, I'm currently testing my proposed solution and set `elfeed-curl-program-name` to the name of a one-line shell script with contents
```
curl "$@" | perl -pe 's/(projecteuclid\.org\/euclid\.e[cj]p\/[0-9]+)_[0-9]+(<\/guid>)/\1\2/g'
```
It works as expected so far and I don't see any problems unless elfeed doesn't only read from stdout/stderr but passes the -o argument to curl (i.e. redirects output to a file), which currently does not seem to be the case. Am I missing something?

**Comment by skeeto** [comment]
Your comment jogged my memory. There's also elfeed-new-entry-parse-hook. 
It runs sooner, and you can override the entry ID. The idea is it lets 
you pluck extra information from the XML and attach it to the entry. In 
your case you could use the XML to figure out a canonical entry ID and 
apply it to the entry.

**Comment by herrsimon** [comment]
That's great! Could you help me out and tell me if a function like this would work (again, my lisp knowledge is extremely limited)?

```ELISP
(defun my/adjust-id (hook type entry db-entry)
  (let ((id (elfeed-entry-id db-entry)))
    (if (function-which-checks-if-id-needs-to-be-adjusted id)
        (setf (elfeed-entry-id db-entry) (some-function-which-does-the-adjustment id))))
```

**Comment by skeeto** [comment]
Yup, you've got the right idea. Here's my own little test hook to make 
sure it works correctly:

(defun example (_type _xml entry)
  (let* ((old-id (elfeed-entry-id entry))
         (new-id (sha1 (cdr old-id))))
    (setf (cdr old-id) new-id)))

Your some-function-which-does-the-adjustment-id would derive it from the 
XML for that entry, such as a digest of a more stable component (title, 
etc.).

Note: An ID is a 2-tuple of the feed's domain and the actual ID given by 
the feed. The reason for the tuple is that, as you know, not every feed 
is well-constructed. There have been feeds that don't choose unique IDs 
for their entries, resulting in collisions between feeds. The domain 
therefore namespaces the IDs.

**Comment by herrsimon** [comment]
Perfect, got it, thanks a lot for your help and also for having programmed elfeed! I'm definitely not only speaking for myself when saying that it makes my daily workflow so much easier and faster.

**Comment by minad** [comment]
Closing, as the issue seems to be resolved.
